packages/sanity/src/core/form/FormBuilderProvider.tsx

```diff
@@ -118,10 +117,7 @@ export function FormBuilderProvider(props: FormBuilderProviderProps) {
   } = props
 
   const documentValueRef = useRef(documentValue)
-
-  useEffect(() => {
-    documentValueRef.current = documentValue
-  }, [documentValue])
+  documentValueRef.current = documentValue
```

Comment:
Why do we not want to set the value of this ref in a `useEffect` anymore?

Reply:
Great question! Counter question: Why do we want to set it in a useEffect in the first place?

The intention behind doing it this way wasn't very clear from reading the code and it very much stands out as an error to me. If I'm not mistaken, idea was that we're keeping it in a ref so that we can look it up at the time the `getDocument()` function is called, to prevent it from becoming part of the render cycle. Since it should always point at the current document, I can't see any reason to wait for it to update it in an effect. But maybe more important: the way it was, would caused child components calling getDocument() in their render to lag one iteration behind.

That said, it doesn't look like we're using `useFormBuilder().__internal.getDocument()` at all anymore (the only prior usage I could find was by the slug input), so I've removed it completely in ada38de750f1e874d2f04f4fbb2e7cd6d5c7906c.

Reply:
Realized my code search skills failed me, and that useFormBuilder().__internal.getDocument() was still in use by the slug input, which made ada38de break. As an alternative I introduced a separate context that provides a `getDocument(path: Path)` function which is stable throughout the life cycle, and can be used to read the current document value in a callback function (e.g. `getNew[Slug]FromSource()`).

This made it possible to remove value entirely from FormBuilderContext, which is a nice little cleanup too. Take a look at c4964ed9992031cfcedc50e7cf1a2115cc72b538